### ISSUE TITLE
Investigating setting curvePoint.y to null for discontinuities.

### ISSUE THREAD

**Issue by veillette** [phetsims/calculus-grapher#133]
In #132 , we discussed the different approaches to keep track of discontinuities in the model.
@pixelzoom volunteered to investigate what would be the effect of setting y values of curvePoint to null or number on  typescript.

**Comment by pixelzoom** [comment]
Quick report on changing CurvePoint to `y: number | null` …

After changing CurvePoint.ts (which was easy), there were 6 other files that reported TS errors:

```
% tsc
...
   4 js/common/model/Curve.ts:214
   4 js/common/model/DerivativeCurve.ts:91
   3 js/common/model/IntegralCurve.ts:68
  18 js/common/model/TransformedCurve.ts:82
   3 js/common/view/CurveNode.ts:201
   3 js/common/view/TransformedCurveNode.ts:156
```

TransformedCurve.ts had the most TS errors, mostly related to point.lastSaveY.  For the moment, I simply added the non-null assertion operator (point.lastSaveY!), and the sim continued to operate without failing, even in the last 2 screens.  (These cases will require more investigation to address robustly.)

The other 5 files were pretty easy to resolve the TS errors, with judicious use of ! and addition of assert  to verify assumptions about when it’s safe to use !. (edited) 

So this direction looks promising!

**Comment by pixelzoom** [comment]
Here's my patch for CurverPoint.ts, in case we need this later.

<details>
<summary>Patch</summary>

```diff
Index: js/common/model/CurvePoint.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/common/model/CurvePoint.ts b/js/common/model/CurvePoint.ts
--- a/js/common/model/CurvePoint.ts	(revision 5bc24bbeeaf04a387bec5470baa12366fb1b4ff7)
+++ b/js/common/model/CurvePoint.ts	(date 1671049725845)
@@ -16,6 +16,7 @@
 
 import Vector2 from '../../../../dot/js/Vector2.js';
 import IOType from '../../../../tandem/js/types/IOType.js';
+import NullableIO from '../../../../tandem/js/types/NullableIO.js';
 import NumberIO from '../../../../tandem/js/types/NumberIO.js';
 import StringIO from '../../../../tandem/js/types/StringIO.js';
 import calculusGrapher from '../../calculusGrapher.js';
@@ -25,15 +26,15 @@
 export type PointType = ( typeof PointTypeValues )[number];
 
 export type PointState = {
-  y: number;
+  y: number | null;
   pointType: PointType;
 };
 
 export type CurvePointStateObject = {
   x: number;
-  y: number;
+  y: number | null;
   pointType: PointType;
-  initialY: number;
+  initialY: number | null;
   initialPointType: PointType;
 };
 
@@ -44,7 +45,7 @@
   // Using an observable Property for the y-value was considered, but it was deemed to be
   // invasive to the performance of the simulation as observers had to listen to the yProperty
   // of all CurvePoints. See https://github.com/phetsims/calculus-grapher/issues/19
-  public y: number;
+  public y: number | null;
 
   public pointType: PointType;
 
@@ -54,7 +55,7 @@
   // an array of all of this Point's saved states.
   private readonly savedStates: PointState [];
 
-  public constructor( x: number, y = 0, pointType: PointType = 'smooth' ) {
+  public constructor( x: number, y: number | null = 0, pointType: PointType = 'smooth' ) {
     assert && assert( Number.isFinite( x ) && CalculusGrapherConstants.CURVE_X_RANGE.contains( x ), `invalid x: ${x}` );
 
     this.x = x;
@@ -69,8 +70,8 @@
     this.savedStates = [];
   }
 
-  public toVector(): Vector2 {
-    return new Vector2( this.x, this.y );
+  public toVector(): Vector2 | null {
+    return ( this.y === null ) ? null : new Vector2( this.x, this.y );
   }
 
   /**
@@ -93,7 +94,7 @@
   /**
    * Gets the most recently saved y-value.
    */
-  public get lastSavedY(): number {
+  public get lastSavedY(): number | null {
     return ( this.savedStates.length === 0 ) ? this.initialState.y : _.last( this.savedStates )!.y;
   }
 
@@ -173,9 +174,9 @@
     valueType: CurvePoint,
     stateSchema: {
       x: NumberIO,
-      y: NumberIO,
+      y: NullableIO( NumberIO ),
       pointType: StringIO,
-      initialY: NumberIO,
+      initialY: NullableIO( NumberIO ),
       initialPointType: StringIO
     },
     toStateObject: ( curvePoint: CurvePoint ) => curvePoint.toStateObject(),
```

</details>

**Comment by veillette** [comment]
Given the result of the discussion in #132, I will assign @pixelzoom to continue his investigation of allowing curvePoint.y to take null for discontinuities.

The first step is probably just to resolve all the TS errors associated with this change (and the ancillaryTool)

The second step would be to set the y value to null somewhere in the code:
from
````
        if ( jump >= 2 ) {
          point.pointType = 'discontinuous';
        }
````
to
````
        if ( jump >= 2 ) {
          point.pointType = 'discontinuous';
          point.y = null;
        }
````

and see that it can pass CT.

**Comment by pixelzoom** [comment]
Over in #132, we decided to proceed with `y: number | null`:

> (1) Use `y: number | null` everywhere, where `null` indicates a discontinuity. This is the recommended option, as long as it meets @amanda-phet's expectations.

So I took at converting, and failed spectacularly. I got into some situations that I did not know how to resolve. I think it would be best if @veillette and I paired on this.

Below is my current patch for CurvePoint.ts.  It's very similar to the patch in https://github.com/phetsims/calculus-grapher/issues/133#issuecomment-1352372788, but I defined:

```typescript
export type CurvePointY = number | null;
```

... instead of duplicating `number | null` in multiple places.

<details>
<summary>patch</summary>

```diff
Index: js/common/model/CurvePoint.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/common/model/CurvePoint.ts b/js/common/model/CurvePoint.ts
--- a/js/common/model/CurvePoint.ts	(revision 4ae56e7468174baeaea1766dd86e47b561119059)
+++ b/js/common/model/CurvePoint.ts	(date 1673313434414)
@@ -16,6 +16,7 @@
 
 import Vector2 from '../../../../dot/js/Vector2.js';
 import IOType from '../../../../tandem/js/types/IOType.js';
+import NullableIO from '../../../../tandem/js/types/NullableIO.js';
 import NumberIO from '../../../../tandem/js/types/NumberIO.js';
 import StringUnionIO from '../../../../tandem/js/types/StringUnionIO.js';
 import calculusGrapher from '../../calculusGrapher.js';
@@ -25,15 +26,19 @@
 export type PointType = ( typeof PointTypeValues )[number];
 
 export type PointState = {
-  y: number;
+  y: CurvePointY;
   pointType: PointType;
 };
 
+// y value for CurvePoint, null when there is a discontinuity.
+// See https://github.com/phetsims/calculus-grapher/issues/132
+export type CurvePointY = number | null;
+
 export type CurvePointStateObject = {
   x: number;
-  y: number;
+  y: CurvePointY;
   pointType: PointType;
-  initialY: number;
+  initialY: CurvePointY;
   initialPointType: PointType;
 };
 
@@ -44,7 +49,7 @@
   // Using an observable Property for the y-value was considered, but it was deemed to be
   // invasive to the performance of the simulation as observers had to listen to the yProperty
   // of all CurvePoints. See https://github.com/phetsims/calculus-grapher/issues/19
-  public y: number;
+  public y: CurvePointY;
 
   public pointType: PointType;
 
@@ -54,7 +59,7 @@
   // an array of all of this Point's saved states.
   private readonly savedStates: PointState [];
 
-  public constructor( x: number, y = 0, pointType: PointType = 'smooth' ) {
+  public constructor( x: number, y: CurvePointY = 0, pointType: PointType = 'smooth' ) {
     assert && assert( Number.isFinite( x ) && CalculusGrapherConstants.CURVE_X_RANGE.contains( x ), `invalid x: ${x}` );
 
     this.x = x;
@@ -69,8 +74,8 @@
     this.savedStates = [];
   }
 
-  public toVector(): Vector2 {
-    return new Vector2( this.x, this.y );
+  public toVector(): Vector2 | null {
+    return ( this.y === null ) ? null : new Vector2( this.x, this.y );
   }
 
   /**
@@ -93,7 +98,7 @@
   /**
    * Gets the most recently saved y-value.
    */
-  public get lastSavedY(): number {
+  public get lastSavedY(): CurvePointY {
     return ( this.savedStates.length === 0 ) ? this.initialState.y : _.last( this.savedStates )!.y;
   }
 
@@ -173,9 +178,9 @@
     valueType: CurvePoint,
     stateSchema: {
       x: NumberIO,
-      y: NumberIO,
+      y: NullableIO( NumberIO ),
       pointType: StringUnionIO( PointTypeValues ),
-      initialY: NumberIO,
+      initialY: NullableIO( NumberIO ),
       initialPointType: StringUnionIO( PointTypeValues )
     },
     toStateObject: ( curvePoint: CurvePoint ) => curvePoint.toStateObject(),
```

</details>

**Comment by pixelzoom** [comment]
@veillette will document our choices here and close.  Probably similar information to https://github.com/phetsims/calculus-grapher/issues/132.

**Comment by veillette** [comment]
Discussed at the 1/10 meeting. The conclusion was that it proves to be problematic to set curvePoint.y to null.  There are many cases in TransformedCurve that assumes that curvePoint.y is a number so that makes it cumbersome to change. 
Instead we opted to have the underlying model to set curvePoint.y to number but to store the information about the discontinuity in the `PointType` that can tajkkes the following values.
`
const PointTypeValues = [ 'smooth', 'cusp', 'discontinuous' ] as const;
`

For the ancillary tools, where we store, for a given x- position, the y value associated with the four curves, it is not possible for the PHET-IO designer to know that a particular x value might be associated with a discontinuity. To remedy to this problem, we opted to create a PhET-IO only property that can return null if the point is at a discontinuity 

```js
  // We are creating this Property solely for PhET-iO, to present the y value. For discontinuities, we present null.
  // This Property persists because it is referenced by the tandem registry.
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const yPhetioProperty = new DerivedProperty( [ curvePointProperty ],
    curvePoint => curvePoint.isDiscontinuous ? null : curvePoint.y, {
      tandem: tandem,
      phetioValueType: NullableIO( NumberIO )
    } );
```

This seems like a nice compromise given the difficulties associated the underlying curve model.
